### ISSUE TITLE
error CERTIFICATE_VERIFY_FAILED

### ISSUE THREAD

**Issue by bieniu** [sivel/speedtest-cli#568]
Two days ago, speedtest-cli 2.0.2 stopped working for me and it displays the errors:
```
maciek@router:/tmp/home/root# speedtest-cli --simple --no-pre-allocate
Cannot retrieve speedtest configuration
ERROR: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] unknown error: unable to get local issuer certificate (_ssl.c:1051)>
maciek@router:/tmp/home/root# speedtest-cli
Retrieving speedtest.net configuration...
Cannot retrieve speedtest configuration
ERROR: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] unknown error: unable to get local issuer certificate (_ssl.c:1051)>
```
Environment: AsusWRT Entware-ng, Python 2.7.15

**Comment by wheelq** [comment]
Same here, works on other devices, except one (which was also working till 19th of Jan)

**Comment by 0800-grizzly** [comment]
Hi,

My speedtest-cli stopped working for the same reason on Jan 25, in the evening. This on a QNAP Entware.

rgds,

Björn

**Comment by wheelq** [comment]
FIX

TL;DR:
run:
`export PYTHONHTTPSVERIFY=0`

Explanation:
LB is redirecting to the HTTPS even if you use HTTP:
`...
Location: https://www.speedtest.net/speedtest-config.php [following]
Spider mode enabled. Check if remote file exists.
--2019-01-22 19:11:45--  https://www.speedtest.net/speedtest-config.php
Connecting to www.speedtest.net|151.101.2.219|:443... connected.
ERROR: cannot verify www.speedtest.net's certificate, issued by 'CN=GlobalSign CloudSSL CA - SHA256 - G3,O=GlobalSign nv-sa,C=BE':
  Unable to locally verify the issuer's authority.
To connect to www.speedtest.net insecurely, use --no-check-certificate'.
...`

You can turn certificate verification off by running the following command:
`export PYTHONHTTPSVERIFY=0`

Issue can be closed

**Comment by bieniu** [comment]
@wheelq Unfortunately, your solution doesn't work for me:
```
maciek@router:/jffs/scripts# export PYTHONHTTPSVERIFY=0
maciek@router:/jffs/scripts# speedtest-cli
Retrieving speedtest.net configuration...
Cannot retrieve speedtest configuration
ERROR: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] unknown error: unable to get local issuer certificate (_ssl.c:1051)>
```

**Comment by 0800-grizzly** [comment]
> FIX
> 
> TL;DR:
> run:
> `export PYTHONHTTPSVERIFY=0`

@wheelq, thanks, this worked for me!

rgds,

Björn

**Comment by wheelq** [comment]
> @wheelq Unfortunately, your solution doesn't work for me:
> 
> ```
> maciek@router:/jffs/scripts# export PYTHONHTTPSVERIFY=0
> maciek@router:/jffs/scripts# speedtest-cli
> Retrieving speedtest.net configuration...
> Cannot retrieve speedtest configuration
> ERROR: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] unknown error: unable to get local issuer certificate (_ssl.c:1051)>
> ```

after this please run:
`echo $PYTHONHTTPSVERIFY` and provide the results (here or via email)

**Comment by bieniu** [comment]
@wheelq 
```
maciek@router:/jffs/scripts# export PYTHONHTTPSVERIFY=0
maciek@router:/jffs/scripts# echo $PYTHONHTTPSVERIFY
0
maciek@router:/jffs/scripts# speedtest-cli
Retrieving speedtest.net configuration...
Cannot retrieve speedtest configuration
ERROR: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] unknown error: unable to get local issuer certificate (_ssl.c:1051)>
maciek@router:/jffs/scripts# echo $PYTHONHTTPSVERIFY
0
```

**Comment by wheelq** [comment]
@bieniu please provide the results for:

`wget -S --spider speedtest.net/speedtest-config.php
`

**Comment by bieniu** [comment]
@wheelq 
```
maciek@router:/jffs/scripts# wget -S --spider speedtest.net/speedtest-config.php
Spider mode enabled. Check if remote file exists.
--2019-01-28 11:16:32--  http://speedtest.net/speedtest-config.php
Resolving speedtest.net... 151.101.66.219, 151.101.130.219, 151.101.194.219, ...
Connecting to speedtest.net|151.101.66.219|:80... connected.
HTTP request sent, awaiting response...
  HTTP/1.1 301 Moved Permanently
  Server: Varnish
  Retry-After: 0
  Content-Type:
  Content-Length: 0
  Accept-Ranges: bytes
  Date: Mon, 28 Jan 2019 10:16:32 GMT
  Via: 1.1 varnish
  Connection: close
  X-Served-By: cache-hhn1530-HHN
  X-Cache: MISS
  X-Cache-Hits: 0
  X-Timer: S1548670593.831295,VS0,VE0
  location: https://www.speedtest.net/speedtest-config.php
Location: https://www.speedtest.net/speedtest-config.php [following]
Spider mode enabled. Check if remote file exists.
--2019-01-28 11:16:32--  https://www.speedtest.net/speedtest-config.php
Resolving www.speedtest.net... 151.101.114.219
Connecting to www.speedtest.net|151.101.114.219|:443... connected.
HTTP request sent, awaiting response...
  HTTP/1.1 200 OK
  Content-Type: application/xml
  Content-Length: 13644
  Date: Mon, 28 Jan 2019 10:16:33 GMT
  Connection: keep-alive
  Accept-Ranges: bytes
  Vary: Accept-Encoding
Length: 13644 (13K) [application/xml]
Remote file exists.
```

**Comment by bieniu** [comment]
I had to install the ca-certificates package and now speedtest-cli works. Thanks @wheelq

**Comment by wheelq** [comment]
@bieniu 
1. From the log looks like you're being routed to the non-SSL page, which is good, but for some reason script tried connecting using SSL...
2. Please share with the rest of us how did you install the missing certificate as this might be another way for fixing this.

Thanks

**Comment by bieniu** [comment]
@wheelq 
```
opkg install ca-certificates
```

**Comment by jupzup** [comment]
I had the same issue on my dreambox with enigma2 but both "opkg ca update" and "export PYTHONHTTPSVERIFY=0" did not solve the problem. 

This was succesfully: Type in the speedtest.py following commands

```
import ssl
ssl._create_default_https_context = ssl._create_unverified_context

```
After that the script will run without issues.

**Comment by Chadster766** [comment]
In Debian Stretch the apt-get speedtest-cli package doesn't work but I i run the below commands after package install it does:

```
wget --no-check-certificate -O /usr/bin/speedtest-cli https://raw.githubusercontent.com/sivel/speedtest-cli/master/speedtest.py

apt-get install ca-certificates

```

Is it possible for the package maintainer to have Debian update the speedtest-cli apt-get package with the latest speedtest-cli py script and add the ca-certificates as a prerequisite?

**Comment by Cryental** [comment]
> In Debian Stretch the apt-get speedtest-cli package doesn't work but if i run the below commands after package install it does:
> 
> ```
> wget --no-check-certificate -O /usr/bin/speedtest-cli https://raw.githubusercontent.com/sivel/speedtest-cli/master/speedtest.py
> 
> apt-get install ca-certificates
> ```
> Is it possible for the package maintainer to have Debian update the speedtest-cli apt-get package with the latest speedtest-cli py script and add the ca-certificates as a prerequisite?

This solution is working, thank you!

**Comment by Cryental** [comment]
> ```
> wget --no-check-certificate -O /usr/bin/speedtest-cli https://raw.githubusercontent.com/sivel/speedtest-cli/master/speedtest.py
> 
> apt-get install ca-certificates
> ```

This solution is working, thank you!

**Comment by ramses147** [comment]
Easiest solution: https://github.com/sivel/speedtest-cli/pull/586

**Comment by joekerna** [comment]
For me none of the solutions above work.

I recently updated openssl to to 1.1.1d and the script keeps throwing the above error.

**Comment by sinusphi** [comment]
>For me none of the solutions above work.
>
>I recently updated openssl to to 1.1.1d and the script keeps throwing the above error.

@joekerna 

Same here. I tried the solutions above, but none of them fixed the issue.

**Comment by joekerna** [comment]
@sivel Please reopen the issue
Thanks!
